src/services/daprDashboardProvider.ts

```
@@ -0,0 +1,65 @@
+import { getLocalizationPathForFile } from '../util/localization';
+import * as tcp from 'tcp-port-used'
+
+const localize = nls.loadMessageBundle(getLocalizationPathForFile(__filename));
+
+
+export interface DaprDashboardProvider {
+    startDashboard(): Promise<string>;
+}
+
+export default class ProcessBasedDaprDashboardProvider implements DaprDashboardProvider {
+    private port: number | undefined;
+    private openPort: number | undefined
+    constructor(private readonly daprPathProvider: () => string) {
+    }
+
+
+    async startDashboard(): Promise<string> {
+        if(this.port !== undefined) {
+            return `http://localhost:${this.port}`
+        }
+        else {
+            await this.spawnDashboardInstance();
+
+            if(this.openPort !== undefined) {
+                await tcp.waitForStatus(this.openPort, 'localhost', true, 500, 4000).then(() => {
+                    this.port = this.openPort; //store port value if localhost port connection is successful
+                }, (err) => {
+                    // eslint-disable-next-line @typescript-eslint/restrict-plus-operands
+                    const msg = err + localize('dashboard.startDashboard.statusTimeout', ': unable to connect to localhost port ') + this.openPort?.toString();
+                    throw new Error(msg)
+                });
+            }
+            // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
+            return `http://localhost:${this.port}`
+        }
+    }
+
+    async spawnDashboardInstance(): Promise<void>{
+        await this.processSpawn();
+    }
+
+    async findOpenPort(): Promise<number> {
+        return await portfinder.getPortPromise().catch(() => {
+            throw new Error(localize('dashboard.findOpenPort.noOpenPort', 'No open port found.'))
+        });
+    }
+
+    async processSpawn() : Promise<void> {
+        this.openPort = await this.findOpenPort();
+        if(this.openPort !== undefined) {
```

Nit: `findOpenPort()` should only return a port (or throw), so checking for undefined shouldn't be necessary.  (Removal of the `then()` and `catch()` may help make that more clear.)